### PR TITLE
Pipeline, Alina Aisina - 21FPL1

### PR DESCRIPTION
**Pull request by itsaisina**

### CODE REVIEW

File: lab_5_scrapper/target_score.txt [hunk 1]
```diff
@@ -1 +1 @@
-6
\ No newline at end of file
+10
```
Comment (demid5111) [inline lab_5_scrapper/target_score.txt:1]: do you plan to increase your mark?

File: lab_5_scrapper/scrapper_config.json [hunk 2]
```diff
@@ -1,9 +1,39 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": [
+        "https://chelny-izvest.ru/news/widget/list/poslednie-novosti-news-page/page/2",
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json]: if you want to demonstrate recursive crawler, leave only one seed - domain name

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -90,25 +206,47 @@ def __init__(self, config: Config) -> None:
...
 
     def find_articles(self) -> None:
         """
         Finds articles
         """
-        pass
+        for seed_url in self._seed_urls:
+            response = make_request(seed_url, self._config)
+            article_bs = BeautifulSoup(response.text, 'lxml')
+            if response.status_code == 200:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: too nested

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        author_elem = article_soup.find('a', {'class': 'page-main__publish-author'})
+        self.article.author = [elem.get_text(strip=True) for elem in author_elem] \
+            if author_elem else ["NOT FOUND"]
+
+        date_elem = article_soup.find('a', {'class': 'page-main__publish-date'})
+        date_str = date_elem.get_text(strip=True) if date_elem else "NOT FOUND"
+        self.article.date = self.unify_date_format(date_str)
+
+        topic_elem = article_soup.find_all('a', {'class': 'panel-group__title'})[1]
+        self.article.topics = topic_elem.get_text(strip=True) if topic_elem else "NOT FOUND"
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: topics should be a list

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+        title_elem = article_soup.find('h1', {'class': 'page-main__head'})
+        self.article.title = title_elem.get_text(strip=True) if title_elem else "NOT FOUND"
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: not found title?

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
         """
         Finds meta information of article
         """
-        pass
+        author_elem = article_soup.find('a', {'class': 'page-main__publish-author'})
+        self.article.author = [elem.get_text(strip=True) for elem in author_elem] \
+            if author_elem else ["NOT FOUND"]
+
+        date_elem = article_soup.find('a', {'class': 'page-main__publish-date'})
+        date_str = date_elem.get_text(strip=True) if date_elem else "NOT FOUND"
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: no date??

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
-        pass
+        response = make_request(self.full_url, self.config)
+        article_soup = BeautifulSoup(response.text, 'lxml')
+        self._fill_article_with_text(article_soup)
+        self._fill_article_with_meta_information(article_soup)
+        return self.article
 
 
-def prepare_environment(base_path: Union[Path, str]) -> None:
+def prepare_environment(base_path: Union[Path, str], recursion: bool = False) -> None:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: changing API is rejected. discard

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        self.num_visited_urls = 0
+        self.last_file_index = 0
+        self.visited_urls = []
+        self.urls = []
+
+    def _handle_crawler_data(self, mode: str) -> None:
+        """
+        Handles saving, loading, and updating crawler data
+        """
+        if self.crawler_data_path.exists() or mode == 'w':
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: split two absolutely different from conceptual pov things: dump and load

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+            with open(self.crawler_data_path, mode, encoding='utf-8') as file:
+                if mode == 'w':
+                    json.dump(self._get_crawler_data(), file,
+                              ensure_ascii=True, indent=4, separators=(', ', ': '))
+                elif mode == 'r':
+                    self._set_crawler_data(json.load(file))
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: no check for corrupted dump file?

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        """
+        Sets the crawler's data from a dictionary
+        """
+        self.last_file_index = data['last_file_idx']
+        self.num_visited_urls = data['num_visited_urls']
+        self.start_url = data['start_url']
+        self.urls = data['urls']
+        self.visited_urls = data['visited_urls']
+
+    def _save_crawler_data(self) -> None:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: how is it different to `update_file_index(self) `? 😄

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        recursive_crawler.update_file_index()
+        current_url = recursive_crawler.urls[index - 1]
+        parser = HTMLParser(full_url=current_url,
+                            article_id=index,
+                            config=configuration)
+        parsed_article = parser.parse()
+        if isinstance(parsed_article, Article):
+            to_raw(parsed_article)
+            to_meta(parsed_article)
+    recursive_crawler.delete_crawler_data()
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: why?? maybe next time I run, I would like to re-use data from previous runs

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+
+
+def main_recursive() -> None:
+    """
+    Entrypoint for scrapper module
+    """
+    configuration = Config(path_to_config=CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH, recursion=True)
+    recursive_crawler = CrawlerRecursive(config=configuration)
+    if not recursive_crawler.crawler_data_path.exists():
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: should happen in prepare environment

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        recursive_crawler.last_file_index = 1
+        shutil.rmtree(ASSETS_PATH)
+        ASSETS_PATH.mkdir(parents=True)
+    else:
+        recursive_crawler.load_crawler_data()
+    recursive_crawler.find_articles()
+    for index in range(recursive_crawler.last_file_index,
+                       len(recursive_crawler.urls) + 1):
+        recursive_crawler.last_file_index = index
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: your code for recursive crawler should not impact code for parsing articles.

namely, last loop over collected urls an parsing them to Article instances should be the same between `main()` and `main_recursive()`

File: lab_6_pipeline/pos_frequency_pipeline.py [hunk 14]
```diff
@@ -1,25 +1,64 @@
...
-
     Example:
     '2	произошло	происходить	VERB	_	Gender=Neut|Number=Sing|Tense=Past	0	root	_	_'
     """
+    params = token_line.split('\t')
+
+    token = ConlluToken(text=params[1])
+    token.set_position(position=int(params[0]))
+    morph_params = MorphologicalTokenDTO(lemma=params[2],
+                                         pos=params[3])
```
Comment (demid5111) [inline lab_6_pipeline/pos_frequency_pipeline.py]: you lost tags, why?

File: lab_6_pipeline/pos_frequency_pipeline.py [hunk 15]
```diff
@@ -1,25 +1,64 @@
...
 
+from core_utils.article.article import Article, ArtifactType, get_article_id_from_filepath
+from core_utils.article.io import to_meta, from_meta
+from core_utils.article.ud import extract_sentences_from_raw_conllu
+from core_utils.constants import ASSETS_PATH
+from core_utils.visualizer import visualize
+from lab_6_pipeline.pipeline import ConlluToken, ConlluSentence, CorpusManager, MorphologicalTokenDTO
+
 
 def from_conllu(path: Path, article: Optional[Article] = None) -> Article:
```
Comment (demid5111) [inline lab_6_pipeline/pos_frequency_pipeline.py:16]: very nice 🎖️

File: lab_6_pipeline/data/mystem_tags_mapping.json [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        "COM": "ADJ",
+        "CONJ": "CCONJ",
+        "INTJ": "INTJ",
+        "NUM": "NUM",
+        "PART": "PART",
+        "PR": "ADP",
+        "S": "NOUN",
+        "SPRO": "PRON",
+        "V": "VERB",
+        "PRED": "ADV"
```
Comment (demid5111) [inline lab_6_pipeline/data/mystem_tags_mapping.json]: where from do you get it? point to official docs, please

File: lab_6_pipeline/data/mystem_tags_mapping.json [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        "им": "Nom",
+        "род": "Gen",
+        "дат": "Dat",
+        "вин": "Acc",
+        "твор": "Ins",
+        "пр": "Loc",
+        "парт": "Par",
+        "местн": "Loc",
+        "зват": "Voc",
+        "им-acc": "Acc"
```
Comment (demid5111) [inline lab_6_pipeline/data/mystem_tags_mapping.json]: same

File: lab_6_pipeline/data/opencorpora_tags_mapping.json [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        "femn": "Fem",
+        "neut": "Neut"
+    },
+    "Animacy": {
+        "anim": "Anim",
+        "inan": "Inan"
+    },
+    "Tense": {
+        "pres": "Pres",
+        "futr": "Fut",
```
Comment (demid5111) [inline lab_6_pipeline/data/opencorpora_tags_mapping.json]: ?

File: lab_6_pipeline/pipeline.py [hunk 19]
```diff
@@ -18,21 +43,58 @@ def __init__(self, path_to_raw_txt_data: Path):
...
+            raise FileNotFoundError
+
+        if not self.path_to_raw_txt_data.is_dir():
+            raise NotADirectoryError
+
+        file_paths = list(self.path_to_raw_txt_data.glob("*"))
+        if not file_paths:
+            raise EmptyDirectoryError
+
+        raw_files = [file for file in file_paths if file.suffix == ".txt" and file.stem.endswith("_raw")]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: use glob for better pattern definition

File: lab_6_pipeline/pipeline.py [hunk 20]
```diff
@@ -18,21 +43,58 @@ def __init__(self, path_to_raw_txt_data: Path):
...
+        meta_files = [file for file in file_paths if file.suffix == ".json" and file.stem.endswith("_meta")]
+
+        if len(raw_files) != len(meta_files):
+            raise InconsistentDatasetError
+
+        raw_ind = sorted([int(file.stem.split("_")[0]) for file in raw_files])
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:72]: where is check for meta files

File: lab_6_pipeline/pipeline.py [hunk 21]
```diff
@@ -18,21 +43,58 @@ def __init__(self, path_to_raw_txt_data: Path):
...
+        for ind, file_id in enumerate(raw_ind[:-1]):
+            if raw_ind[ind + 1] - file_id > 1:
+                raise InconsistentDatasetError
+
+        for raw_file in raw_files:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: same about meta files

File: lab_6_pipeline/pipeline.py [hunk 22]
```diff
@@ -18,21 +43,58 @@ def __init__(self, path_to_raw_txt_data: Path):
...
+            if raw_file.stat().st_size == 0:
+                raise InconsistentDatasetError
 
     def _scan_dataset(self) -> None:
         """
         Register each dataset entry
         """
+        for raw_file in self.path_to_raw_txt_data.glob("*_raw.txt"):
+            article_id = int(raw_file.stem.split("_")[0])
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:88]: use `from_raw` to directly create Article's instance from path

File: lab_6_pipeline/pipeline.py [hunk 23]
```diff
@@ -55,26 +120,48 @@ def __init__(self, text: str):
...
+        self.position = 0
+        self._morphological_parameters = MorphologicalTokenDTO()
 
     def set_morphological_parameters(self, parameters: MorphologicalTokenDTO) -> None:
         """
         Stores the morphological parameters
         """
+        self._morphological_parameters = parameters
+
+    def set_position(self, position: int) -> None:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: it is public, no need in getter, rm

File: lab_6_pipeline/pipeline.py [hunk 24]
```diff
@@ -55,26 +120,48 @@ def __init__(self, text: str):
...
+        deps = '_'
+        misc = '_'
+
+        return '\t'.join([position, text, lemma, pos, xpos, feats, head, deprel, deps, misc])
 
     def get_cleaned(self) -> str:
         """
         Returns lowercase original form of a token
         """
+        return self._text.lower().translate(str.maketrans('', '', string.punctuation + "«»–"))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: use substituion via regex

File: lab_6_pipeline/pipeline.py [hunk 25]
```diff
@@ -86,21 +173,37 @@ def __init__(self, position: int, text: str, tokens: list[ConlluToken]):
...
+        text = f"# text = {self._text}\n"
+        tokens = f"{self._format_tokens(include_morphological_tags)}\n"
+
+        return f"{sent_id}{text}{tokens}"
 
     def get_cleaned_sentence(self) -> str:
         """
         Returns the lowercase representation of the sentence
         """
+        return " ".join(token.get_cleaned() for token in self._tokens).replace('  ', ' ').strip()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: you could just check if token is not empty when making a generator

File: lab_6_pipeline/pipeline.py [hunk 26]
```diff
@@ -112,11 +215,37 @@ def convert_morphological_tags(self, tags: str) -> str:  # type: ignore
...
+        pos = self.convert_pos(tags)
+        tags = re.sub(r'\(.+?\)', lambda match: match.group(0).split('|')[0], tags)
+        extracted_tags = re.findall(r'[а-я]+', tags)
+        ud_tags = {}
+
+        pos_specific_categories = {
+            "NOUN": [self.case, self.number, self.gender, self.animacy],
+            "VERB": [self.tense, self.number, self.gender],
+            "ADJ": [self.case, self.number, self.gender],
+            "ADV": [],
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: why do we need to create a key for each unsupported pos? rework

File: lab_6_pipeline/pipeline.py [hunk 27]
```diff
@@ -112,11 +215,37 @@ def convert_morphological_tags(self, tags: str) -> str:  # type: ignore
...
+                    ud_tags[category] = self._tag_mapping[category][tag]
+
+        return '|'.join(f'{k}={v}' for k, v in sorted(ud_tags.items()))
 
     def convert_pos(self, tags: str) -> str:  # type: ignore
         """
         Extracts and converts the POS from the Mystem tags into the UD format
         """
+        extracted_pos = re.search(r'[A-Z]+', tags)
+        return self._tag_mapping["POS"].get(extracted_pos[0], "")
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: is it a correct option to return empty string when POS is not recognized?

File: lab_6_pipeline/pipeline.py [hunk 28]
```diff
@@ -128,11 +257,22 @@ def convert_pos(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
         """
         Extracts and converts POS from the OpenCorpora tags into the UD format
         """
+        for oc_tag, ud_tag in self._tag_mapping['POS'].items():
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: you can access pos tag directly from what pymorphy returns

File: lab_6_pipeline/pipeline.py [hunk 29]
```diff
@@ -128,11 +257,22 @@ def convert_pos(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
...
+            if oc_tag in tags:
+                return ud_tag
+        return 'UNKN'
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: you should get this value from mapping, not a magic string defined in a code

File: lab_6_pipeline/pipeline.py [hunk 30]
```diff
@@ -128,11 +257,22 @@ def convert_pos(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
...
 
     def convert_morphological_tags(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
         """
         Converts the OpenCorpora tags into the UD format
         """
+        ud_tags = {}
+        for category in ('Case', 'Number', 'Gender', 'Animacy'):
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: 1. magic constants
2. inspect if these properties can be access directly from OpenCorporaTag

File: lab_6_pipeline/pipeline.py [hunk 31]
```diff
@@ -144,16 +284,61 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        path = Path(__file__).parent / 'data' / 'mystem_tags_mapping.json'
+        self.corpus_manager = corpus_manager
+        self.tag_converter = MystemTagConverter(path)
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        mystem = Mystem()
+        mystem_result = [x for x in mystem.analyze(text) if x['text'].isalnum()]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: pass whole text to mystem

File: lab_6_pipeline/pipeline.py [hunk 32]
```diff
@@ -144,16 +284,61 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        sentences = []
+
+        for sentence_position, sentence in enumerate(split_by_sentence(text)):
+            tokens_with_punctuations = re.findall(r'\w+|[.]', sentence)
+
+            sentence_tokens = []
+            for token_position, token in enumerate(tokens_with_punctuations, start=1):
+                conllu_token = ConlluToken(token)
+
+                if token.isalnum() and result_idx < len(mystem_result):
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: too nested

File: lab_6_pipeline/pipeline.py [hunk 33]
```diff
@@ -165,22 +350,79 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        super().__init__(corpus_manager)
+        self._backup_analyzer = MorphAnalyzer()
+        mapping_path = Path(__file__).parent / 'data' / 'opencorpora_tags_mapping.json'
+        self._backup_tag_converter = OpenCorporaTagConverter(mapping_path)
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        mystem = Mystem()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: apply comments from basic pipeline's `_process`

**Review by demid5111 — CHANGES_REQUESTED** [review]
you have made a lot of work! time to start polishing it

File: lab_6_pipeline/pos_frequency_pipeline.py (hunk 14 above)
Comment (itsaisina) [reply lab_6_pipeline/pos_frequency_pipeline.py]: done

File: lab_6_pipeline/data/mystem_tags_mapping.json (hunk 16 above)
Comment (itsaisina) [reply lab_6_pipeline/data/mystem_tags_mapping.json]: removed

File: lab_6_pipeline/data/mystem_tags_mapping.json (hunk 17 above)
Comment (itsaisina) [reply lab_6_pipeline/data/mystem_tags_mapping.json]: removed

File: lab_6_pipeline/data/opencorpora_tags_mapping.json (hunk 18 above)
Comment (itsaisina) [reply lab_6_pipeline/data/opencorpora_tags_mapping.json]: removed

File: lab_6_pipeline/pipeline.py (hunk 19 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 20 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py:72]: added

File: lab_6_pipeline/pipeline.py (hunk 21 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: added

File: lab_6_pipeline/pipeline.py (hunk 22 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py:88]: done

File: lab_6_pipeline/pipeline.py (hunk 24 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 25 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 26 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: reworked

File: lab_6_pipeline/pipeline.py (hunk 27 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: no, reworked

File: lab_6_pipeline/pipeline.py (hunk 28 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 29 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: reworked

File: lab_6_pipeline/pipeline.py (hunk 30 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: reworked

File: lab_6_pipeline/pipeline.py (hunk 31 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 32 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: reworked

File: lab_6_pipeline/pipeline.py (hunk 33 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 23 above)
Comment (itsaisina) [reply lab_6_pipeline/pipeline.py]: removed

File: lab_5_scrapper/scrapper_config.json (hunk 2 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper_config.json]: ok

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_5_scrapper/scrapper.py (hunk 8 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: removed

File: lab_5_scrapper/scrapper.py (hunk 11 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: removed

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: but it doesn't allow the crawler to keep collecting articles after stopping... it cleans up the environment

File: lab_5_scrapper/scrapper.py (hunk 12 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: Can you please explain to me again the trick of stopping the crawler. 
The way I understand it is that when we stop the crawler, when it starts up again it continues to collect articles from where it stopped. For example, gathered
1_raw.txt
2_raw.txt
3_raw.txt
4_raw.txt
and then I press ctrl c
and then I start again and go on collecting 
5_raw.txt
6_raw.txt
7_raw.txt
8_raw.txt
Same?

At my first attempt to submit, you were not satisfied with this. That the articles began to be reassembled.

Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: That's what I'm trying to implement. 
But now the guys who passed a grade of 10 said that it is not the text files with articles should be collected in this way, namely the links.

File: lab_6_pipeline/data/mystem_tags_mapping.json [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
+    "Animacy": {
+        "од": "Anim",
+        "неод": "Inan"
+    },
+    "Tense": {
+        "наст": "Pres",
+        "непрош": "Fut",
+        "прош": "Past"
+    }
+}
```
Comment (demid5111) [inline lab_6_pipeline/data/mystem_tags_mapping.json]: newline

File: lab_6_pipeline/data/opencorpora_tags_mapping.json [hunk 35]
```diff
@@ -0,0 +1,49 @@
...
+    "Gender": {
+        "masc": "Masc",
+        "femn": "Fem",
+        "neut": "Neut"
+    },
+    "Animacy": {
+        "anim": "Anim",
+        "inan": "Inan"
+    }
+}
```
Comment (demid5111) [inline lab_6_pipeline/data/opencorpora_tags_mapping.json]: newline

File: lab_5_scrapper/target_score.txt (hunk 1 above)
Comment (demid5111) [reply lab_5_scrapper/target_score.txt:1]: ?

File: lab_5_scrapper/scrapper_config.json (hunk 2 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper_config.json]: and? 😆

File: lab_6_pipeline/pipeline.py [hunk 36]
```diff
@@ -128,11 +248,28 @@ def convert_pos(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
...
+        ud_tags = {}
+        tag_properties = {
+            self.case: tags.case,
+            self.number: tags.number,
+            self.gender: tags.gender,
+            self.animacy: tags.animacy,
+        }
+
+        for category, oc_tag in tag_properties.items():
+            if oc_tag is not None:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: too nested

File: lab_6_pipeline/pipeline.py [hunk 37]
```diff
@@ -144,16 +281,64 @@ def __init__(self, corpus_manager: CorpusManager):
...
         """
+        conllu_sentences = []
+        result = (i for i in self._mystem.analyze(text))
+        for sentence_position, sentence in enumerate(split_by_sentence(text)):
+            conllu_tokens = []
+            original_sentence = sentence
+
+            tokens = []
+            for token in result:
+                if token['text'] not in sentence:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: do you suggest for each sentence to go through all text and skip unrelated words? and unrelated means not a duplicate of this word? how do you prove that this word is not a duplication and is exactly from this sentence?

File: lab_6_pipeline/pipeline.py [hunk 38]
```diff
@@ -144,16 +281,64 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        return conllu_sentences
 
     def run(self) -> None:
         """
         Performs basic preprocessing and writes processed text to files
         """
+        for article in self._corpus.get_articles().values():
+            article.set_conllu_sentences(self._process(article.text))
+            to_cleaned(article)
+            to_conllu(article, include_morphological_tags=False, include_pymorphy_tags=False)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:362]: nice!

File: lab_6_pipeline/pipeline.py [hunk 39]
```diff
@@ -165,22 +350,79 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        self._backup_analyzer = MorphAnalyzer()
+        mapping_path = Path(__file__).parent / 'data' / 'opencorpora_tags_mapping.json'
+        self._backup_tag_converter = OpenCorporaTagConverter(mapping_path)
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        conllu_sentences = []
+        result = (i for i in self._mystem.analyze(text))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:386]: looks like duplication of simple pipeline, how to minimize duplication?

File: lab_6_pipeline/pipeline.py [hunk 40]
```diff
@@ -165,22 +350,79 @@ def __init__(self, corpus_manager: CorpusManager):
...
+            tokens.append({'text': '.'})
+
+            for token_position, token in enumerate(tokens, start=1):
+                text = token['text']
+                if 'analysis' in token and token['analysis']:
+                    pos = self._converter.convert_pos(token['analysis'][0]['gr'])
+                    if pos == 'NOUN':
+                        lex = self._backup_analyzer.parse(text)[0].normal_form
+                        open_corpora_tags = self._backup_analyzer.parse(text)[0].tag
+                        pos = self._backup_tag_converter.convert_pos(open_corpora_tags)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: pymorphy converter also accepts output of pymorphy without preprocessing

**Review by demid5111 — CHANGES_REQUESTED** [review]
almost, final effort is needed

File: lab_6_pipeline/data/mystem_tags_mapping.json (hunk 34 above)
Comment (itsaisina) [reply lab_6_pipeline/data/mystem_tags_mapping.json]: added

File: lab_6_pipeline/data/opencorpora_tags_mapping.json (hunk 35 above)
Comment (itsaisina) [reply lab_6_pipeline/data/opencorpora_tags_mapping.json]: added

File: lab_5_scrapper/scrapper_config.json (hunk 2 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper_config.json]: I forgot to push

File: lab_5_scrapper/target_score.txt (hunk 1 above)
Comment (itsaisina) [reply lab_5_scrapper/target_score.txt:1]: yes, at least I'll try to

File: lab_5_scrapper/scrapper.py (hunk 7 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: finally reworked

File: lab_5_scrapper/scrapper.py (hunk 13 above)
Comment (itsaisina) [reply lab_5_scrapper/scrapper.py]: reworked

File: lab_6_pipeline/pipeline.py [hunk 41]
```diff
@@ -112,11 +210,33 @@ def convert_morphological_tags(self, tags: str) -> str:  # type: ignore
...
+                if tag in self._tag_mapping[category]:
+                    ud_tags[category] = self._tag_mapping[category][tag]
+
+        return '|'.join(f'{k}={v}' for k, v in sorted(ud_tags.items()))
 
     def convert_pos(self, tags: str) -> str:  # type: ignore
         """
         Extracts and converts the POS from the Mystem tags into the UD format
         """
+        extracted_pos = re.search(r'[A-Z]+', tags)[0]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:238]: no guess from mystem then what?

File: lab_6_pipeline/pipeline.py [hunk 42]
```diff
@@ -128,11 +248,30 @@ def convert_pos(self, tags: OpencorporaTagProtocol) -> str:  # type: ignore
         """
         Extracts and converts POS from the OpenCorpora tags into the UD format
         """
+        oc_pos = tags.POS
+        return self._tag_mapping['POS'].get(oc_pos, self._tag_mapping['POS'].get('UNKN'))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: magic..

File: lab_6_pipeline/pipeline.py [hunk 43]
```diff
@@ -144,16 +283,80 @@ def __init__(self, corpus_manager: CorpusManager):
...
+        mapping_path = Path(__file__).parent / 'data' / 'mystem_tags_mapping.json'
+        self._converter = MystemTagConverter(mapping_path)
+
+    def _extract_tokens(self, sentence: str, result: Generator) -> List[Dict[str, str]]:
+        """
+        Extracts tokens from the sentence
+        """
+        tokens = []
+        while sentence:
+            token = next(result)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: ? StopIteration never raised? how do you guarantee that?

File: lab_6_pipeline/pipeline.py [hunk 44]
```diff
@@ -144,16 +283,80 @@ def __init__(self, corpus_manager: CorpusManager):
...
+
+            if token['text'] not in sentence:
+                continue
+
+            sentence = sentence.replace(token['text'], '', 1)
+            if any(c.isalnum() for c in token['text']):
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: ?

**Review by demid5111 — CHANGES_REQUESTED** [review]
look into comments

**Pull request closed without merging**